keps/sig-node/xxx-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1408 @@
+  had some ideas what information the scheduler might need (“supports
+  overcommit”, “fractional”), but ultimately any choice regarding that
+  will only work for certain kinds of resources.
+* A standardized way of describing available resources. Only allocated
+  resources are visible through the APIs defined below. How to
+  describe available resources is plugin specific because it depends
+  on the kind of resource which attributes might be relevant. Plugins
+  should use and document their individual approach for this (for
+  example, defining a CRD and publishing through that).
+* An abstraction layer for resource requests, i.e., something like a
+  “I want some kind of GPU”. Users will need to know about specific
+  resource plugins and which parameters they support. Portability of
+  workloads could be added on top of this proposal by introducing the
+  selection of a resource implementation through labels and
+  standardizing those labels and the associated parameters. The
+  [Resource Class
+  Proposal](https://docs.google.com/document/d/1qKiIVs9AMh2Ua5thhtvWqOqW0MSle_RV3lfriO1Aj6U/edit#heading=h.jzfmfdca34kj)
+  included such an approach.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+The proposal is that a plugin handles all operations that are specific
+to the resources managed by that plugin. This includes operations at
+the control plane level (tracking where in the cluster resources are
+available, helping with pod scheduling decisions, allocating resources
+when requested) as well as the node level (preparing container
+startup). Such a plugin can be implemented in arbitrary programming
+languages as long as it supports the resource allocation protocol and
+gRPC interfaces defined in this KEP. An utility package with Go
+support code will be made available to simplify the development of
+such a plugin, but using it will not be required and its API is not
+part of this KEP.
+
+Three new API object types get added in a new API group:
+- ResourcePlugin, not namespaced, with a description of the plugin.
+- ResourceClass, not namespaced, with privileged parameters for
+  multiple resource instances of a certain kind. All these instances
+  are provided by the same resource plugin, which is identified by a
+  field in the class.
+- ResourceClaim, namespaced, with parameters provided by a normal user
+  that describes a resource instance that needs to be allocated. A
+  ResourceClaim contains the usual meta data, a spec and a status. The
+  spec identifies the plugin that handles the resource via a class
+  name.
```

Comment:
I admittedly don't feel qualified to comment on whether this is a good idea or not, but it's definitely very different than what is done today. Can someone comment on whether there is any precedence for something like this (i.e. using a CRD to hold information like this AND have an API to the kubelet for other types of operations)?

Reply:
This is all very similar to how volumes are handled (PersistentVolumeClaim, PersistentVolume). One improvement is that the separate PersistentVolume object gets avoided. The approach for volumes came from avoiding non-observable status in the PersistentVolumeClaim. I think there is consensus that this led to a very complex design. SIG-Storage is still working on fixing bugs with it. Tim's pending KEP where he proposes to allow status to hold non-observable information mentions volumes.